File: sorting-algotithms/quickSort/quick_sort.rb
def pivotHelper(arr, start_idx, end_index)
    pivot = arr[start_idx]
    idx = start_idx

    (start_idx...end_index).each do |i|
        if arr[i] < pivot
            idx += 1
            arr[i], arr[idx] = arr[idx], arr[i]
        end
    end
    arr[start_idx], arr[idx] = arr[idx], arr[start_idx]
    return idx
end

def quick_sort(arr, left, right)
    if left < right
        idx = pivotHelper(arr, left, right)
        quick_sort(arr, left, idx)
        quick_sort(arr, idx + 1, right)
    end
    return arr
end


arr = [5,7,4,1,6,8,7,9,2,3,7]
print quick_sort(arr, 0, arr.size)

File: sorting-algotithms/quickSort/quickSort.js
function pivotHelper(arr, start=0, end=arr.length) {
  const swap = (arr, i, j) => ([arr[i], arr[j]] = [arr[j], arr[i]]);
  let pivot = arr[start];
  let swapIdx = start;

  for (let i = start + 1; i < end; i++) {
    if (pivot > arr[i]) {
      swapIdx++;
      swap(arr, swapIdx, i);
    }
  }
  swap(arr, start, swapIdx);
  return swapIdx;
}

function quickSort(arr, left=0, right=arr.length) {
  if (left < right) {
    let pivotIdx = pivotHelper(arr, left, right);
    quickSort(arr, left, pivotIdx);
    quickSort(arr, pivotIdx + 1, right);
  }
  return arr;
}
